src/editor/components/modals/file-modal.jsx

```diff
@@ -0,0 +1,472 @@
+  static propTypes = {
+    notebookID: PropTypes.number.isRequired,
+    savedFiles: PropTypes.arrayOf(PropTypes.object),
+    maxFilenameLength: PropTypes.number,
+    maxFileSize: PropTypes.number
+  };
+
+  static defaultProps = {
+    savedFiles: [],
+    maxFilenameLength: false,
+    maxFileSize: false
+  };
+
+  constructor(props) {
+    super(props);
+
+    this.fileInputRef = React.createRef();
+
+    // Files are keyed by unique IDs known only to this component. These are not
+    // the same as the IDs that saveFileToServer returns. (As the name implies,
+    // those are only set for saved files.) Setting our own IDs here helps us to
+    // keep track of files before, during, and after upload.
+    const files = {};
+    props.savedFiles.forEach(f => {
+      files[this.getNewFileKey()] = {
+        status: "saved",
+        file: f
+      };
+    });
+
+    this.state = {
+      files,
+      pendingDelete: false
+    };
+  }
+
+  componentWillUnmount() {
+    // We don't want the timeout callback to execute after this component is
+    // unmounted if a user closes the file modal immediately after copying a
+    // path.
+    if (this.copiedFeedbackTimeout) {
+      clearTimeout(this.copiedFeedbackTimeout);
+    }
+  }
+
+  /**
+   * Return a unique key for keeping track of a file, starting with 0 and
+   * incrementing by one each additional time this method is called.
+   */
+  getNewFileKey = () => {
+    if (this.lastFileKey === undefined) {
+      this.lastFileKey = 0;
+    } else {
+      this.lastFileKey += 1;
+    }
+    return this.lastFileKey;
+  };
+
+  /**
+   * Upload all files that the user selected.
+   */
+  receiveFiles = e => {
+    Array.from(e.target.files).forEach(unsavedFile => {
+      const key = this.getNewFileKey();
+
+      if (this.props.maxFileSize && unsavedFile.size > this.props.maxFileSize) {
+        this.setState(state => {
+          const newState = state;
+          const maxFileSizeMB = this.props.maxFileSize / 1024 / 1024;
+          newState.files[key] = {
+            status: "error",
+            errorMessage: `File exceeds maximum size of ${maxFileSizeMB}MB`,
+            file: unsavedFile
+          };
+          return newState;
+        });
+      } else if (
+        this.props.maxFilenameLength &&
+        unsavedFile.name.length > this.props.maxFilenameLength
+      ) {
+        this.setState(state => {
+          const newState = state;
+          newState.files[key] = {
+            status: "error",
+            errorMessage: `Filename exceeds maximum length of ${
+              this.props.maxFilenameLength
+            } characters`,
+            file: unsavedFile
+          };
+          return newState;
+        });
+      } else {
+        // Update state immediately to reflect that this file has begun
+        // uploading. This allows us to show a spinner in its row for the
+        // duration of the upload.
+        this.setState(state => {
```

Comment:
For sure. You're right that objects like `state` are passed by reference. I used this approach because it satisfies the *airbnb* style guidelines, but you're right that it satisfies the guidelines in name only. I'll use `Object.assign`.

Reply:
Great tip, by the way. I didn't consider that the `Array.prototype.slice` cloning method doesn't copy inner objects by value. Interesting!